### PR TITLE
[Dev] Remove version linter flakiness

### PR DESCRIPTION
**Pull request by dkijania**
An attempt to fix version linter job flakiness by retrieving more than 1 commit from base and release branch and trying to find reference type shape best on shas retrieved in such way. For pr branch still we are getting only one (last) commit.

Closes: #14812

### CODE REVIEW

**Comment by dkijania** [comment]
!ci-build-me

File: scripts/version-linter.py [hunk 1]
```diff
@@ -24,26 +24,58 @@
...
 
 def set_error():
   global exit_code
   exit_code=1
 
-def branch_commit(branch):
+def branch_commits(branch,n):
   print ('Retrieving', branch, 'head commit...')
-  result=subprocess.run(['git','log','-n','1','--format="%h"','--abbrev=7','--no-merges',f'{branch}'],
+  result=subprocess.run(['git','log','-n',str(n),'--format="%h"','--abbrev=7','--no-merges',f'{branch}'],
```
Comment (mrmr1993) [inline scripts/version-linter.py:37]: I think this is the opposite of what we want. We really want the last n 'first-parent' commits, each of which represents a merge to the target branch from a PR. Consider instead
```
git log -n N --format="%h" --abbrev=7 --first-parent BRANCH
```

**Review by mrmr1993 — CHANGES_REQUESTED** [review]
See comment. I'm pretty sure that fixing that bug will also induce the flakiness again though; it would probably be more disciplined to generate and upload the shapes for the parent commit as required instead, in some job that the check depends upon, and short circuiting when the corresponding shape file exists.

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

File: buildkite/scripts/version-linter-patch-missing-type-shapes.sh [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    git submodule sync
+    git submodule update --init --recursive
+    eval $(opam config env)
+    TYPE_SHAPE_FILE=${__commit:0:7}-type_shape.txt
+    dune exec src/app/cli/src/mina.exe internal dump-type-shapes > /tmp/${TYPE_SHAPE_FILE}
+    revert_checkout
+    source buildkite/scripts/gsutil-upload.sh /tmp/${TYPE_SHAPE_FILE} gs://mina-type-shapes
+}
+
+if ! $(gsutil ls gs://mina-type-shapes/$BUILDKITE_COMMIT 2>/dev/null); then
```
Comment (georgeee) [inline buildkite/scripts/version-linter-patch-missing-type-shapes.sh:37]: `BUILDKITE_COMMIT` and `RELEASE_BRANCH_COMMIT` are treated uniformly here, however they seem to have different semantics.

`RELEASE_BRANCH_COMMIT` is abbreviated to 7 symbols, and hence using it in file name is what one would expect.

On the other hand, `gsutil ls gs://mina-type-shapes/$BUILDKITE_COMMIT` will always fail, because the `BUILDKITE_COMMIT` contains full commit name, AFAIU

I suggest adding code to abbreviate `BUILDKITE_COMMIT` to 7 symbols as well.

After that the code above could be simplified as well:

```
TYPE_SHAPE_FILE=${__commit:0:7}-type_shape.txt
```

to

```
TYPE_SHAPE_FILE=$__commit-type_shape.txt
```

File: buildkite/scripts/version-linter-patch-missing-type-shapes.sh [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    checkout_and_dump $BUILDKITE_COMMIT
+fi
+
+if ! $(gsutil ls gs://mina-type-shapes/$RELEASE_BRANCH_COMMIT 2>/dev/null); then
+    checkout_and_dump $RELEASE_BRANCH_COMMIT
+fi
+
+if [[ -n "${BUILDKITE_PULL_REQUEST_BASE_BRANCH:-}" ]]; then
```
Comment (georgeee) [inline buildkite/scripts/version-linter-patch-missing-type-shapes.sh:45]: Why is this needed?

I see that we have a check for the PR branch in version linter, but do we really need that check?

File: scripts/version-linter.py [hunk 4]
```diff
@@ -250,18 +269,18 @@ def assert_commit(commit, desc):
 
   subprocess.run(['git','fetch'],capture_output=False)
 
-  base_branch_commit=branch_commit(base_branch)
-  download_type_shapes('base',base_branch,base_branch_commit)
+  base_branch_commit = find_latest_type_shape_ref_on(base_branch,n=10)
```
Comment (georgeee) [inline scripts/version-linter.py:272]: Why is this change introduced?

What if we pick the wrong commit, one that actually has a different type shape from the last branch commit?

**Review by georgeee — CHANGES_REQUESTED** [review]

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

File: buildkite/scripts/version-linter-patch-missing-type-shapes.sh (hunk 3 above)
Comment (dkijania) [reply buildkite/scripts/version-linter-patch-missing-type-shapes.sh:45]: If you are running or rerunning pipeline directly from BK this variable is not set. We want to make sure version linter works also when running not through !ci-build-me

File: scripts/version-linter.py (hunk 4 above)
Comment (dkijania) [reply scripts/version-linter.py:272]: We are usually looking in mainline branches for reference so there is a silent agreement that if PR with changes in version linter was merged to compatible or develop it is a correct one.

File: buildkite/scripts/version-linter-patch-missing-type-shapes.sh (hunk 2 above)
Comment (dkijania) [reply buildkite/scripts/version-linter-patch-missing-type-shapes.sh:37]: nice catch !

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

File: scripts/version-linter.py (hunk 4 above)
Comment (georgeee) [reply scripts/version-linter.py:272]: > PR with changes in version linter was merged to compatible or develop it is a correct one

Yes, but what if a PR changed type shape in the mainline branch for good, and we're reverting it here by accident?

It seems to me as if policy to only check against the last commit would be much clearer, inviting less unexpected outcomes so to say.

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
stale. closing
